libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/factory/RouteProgressInstanceFactory.kt

```
@@ -0,0 +1,75 @@
+     * Provides route and location tracking information.
+     * @param currentLegProgress [RouteLegProgress] current progress of the active leg, includes
+     * time and distance estimations.
+     * @param upcomingStepPoints [List][Point] location coordinates describing the upcoming step.
+     * @param inTunnel [Boolean] value indicating whether the current location is in a tunnel.
+     * @param distanceRemaining [Float] provides the distance remaining in meters until the user
+     * reaches the end of the route.
+     * @param distanceTraveled [Float] representing the distance traveled along the route in meters.
+     * @param durationRemaining [Double] seconds time remaining until the route destination is reached.
+     * @param fractionTraveled [Float] fraction traveled along the current route. This value is
+     * between 0 and 1 and isn't guaranteed to reach 1 before the user reaches the end of the route.
+     * @param remainingWaypoints [Int] number of waypoints remaining on the current route.
+     * @param upcomingRoadObjects list of upcoming road objects.
+     * @param stale `true` if there were no location updates for a significant amount which causes
+     * a lack of confidence in the progress updates being sent.
+     */
+    fun buildRouteProgressObject(
+        route: DirectionsRoute,
+        bannerInstructions: BannerInstructions? = null,
+        voiceInstructions: VoiceInstructions? = null,
+        currentState: RouteProgressState = RouteProgressState.INITIALIZED,
+        currentLegProgress: RouteLegProgress? = null,
+        upcomingStepPoints: List<Point>? = null,
+        inTunnel: Boolean = false,
+        distanceRemaining: Float = 0f,
+        distanceTraveled: Float = 0f,
+        durationRemaining: Double = 0.0,
+        fractionTraveled: Float = 0f,
+        remainingWaypoints: Int = 0,
+        upcomingRoadObjects: List<UpcomingRoadObject> = emptyList(),
+        stale: Boolean = false
```

We should take advantage of this being a factory now where we _don't have to define placeholder default values_ and only set true values when constructing the object.

This will also ensure that we're not missing an argument and that it doesn't fall back to a placeholder default value.

---

If you look at the existing implementation it makes using a single constructor or factory method complicated because the current code scatters builder calls though out different `ifNonNull` blocks.  The builder in this case is a cleaner approach IMO. I think we should stick with the existing builder for this class.

---

We can stick with the builder if we think it's ergonomically better for us internally :+1:, but that builder should remain internal (or in an internal package) and not set any default values that we might mistakenly not apply during a building flow. Those cases should be dismissed as errors right away.